types/node-vault-client/index.d.ts

```diff
@@ -0,0 +1,162 @@
+// eslint-disable-next-line @typescript-eslint/naming-convention
+interface IAMAuthConfig {
+    role: string;
+    credentials?: {
+        accessKeyId: string;
+        secretAccessKey: string;
+        sessionToken?: string;
+    };
+    iam_server_id_header_value?: string;
+}
+
+interface AppRoleAuthConfig {
+    role_id: string;
+    secret_id?: string;
```

Comment:
For any of these options bags with optional props, you probably want to be adding `| undefined` to them to support `exactOptionalPropertyTypes`.

See: [Playground Link](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true&ts=4.4.2#code/PTAEHUEsBcAtQKYA8CGBjaB5ADtSB7AOxQBsAFAJ32wQugE8AVemgZ0WICMSEATAGlDMaAZTQVIuALAAoEKFgp2raBIy1QFAK492KKlsK9QAd0XRQ9fFtBoUhUKwQX7oaniKlZ87FRp16UGh8ADpZbzAAMXwKRFQAW2weQThIdkhCaFoAM3QEUF4ENBJ9BHY4FAs4Wny00BQI0F9qWgZTWEg0eDsHTnyifuygk3xHVQyAc1YALlAAcl59AGs5t1i5kkgJ2Gg5xpjQGEdYaxJjQnwLPsOHardOACsi6DCZWQysily0fIBVJwoABEELkdNB2ABvWSgUDyRiwfIoThOQg-NxDFCgABupC0+QoCF8ZQQmXYc1Y9BUCHiexkMLQ+BIMXh1IQmCxtAkhQA-LMAESLChLPmgAA+oD5m220D5AG5ZABfcJyMBQODWKoddjZEoTDhIngCIIIgn1U1wAn5HEkPHsMydeBWGw9RpOCwAAwZTIoLPibI5FC5CHdQVGfX5guFgklWx2IvsxndhkK2QyfHdr0aImceEIeru1rxodASaMILTvBDJkgJBI9Vr+BMoHi+BUoAAUigcWIJLhNIY8H7Gl0ikttQc7sg0llUflgqBcjXBPauod2CDss9IBySIFciQKa95AAJRsIAMpE0Icn1BdaABe97TrHKo1S7D6iixBAoISECIKctCBgAhCEabIDk9RlmQRP12U5SBChDAB9Xh8DKQg5mgZCbmKLRChLZMgPTTMVX-fIdRQPV3WQdAsFwUDSEoFoAmEMoQ3iFAljKIQWAQHtJBcNA0C0ChKgQXdGk46BVzuFMMhAogmiob9CmMJR6gcdxGLrZp-AYUjCmKUo71RDwHAmZx-lobNoFzKYAAoAEpZmsoEQRQMFWHlN4ZAZQg2zdez2AAXlASzoDc2zgucnygsmVgQi9GDWXgwNEPyMKBWWOVZHivNEuSn1YP9BCCKyqU4x8xoACEtAsC5oGmPKcwSpLoOK1KAyDUAwqI1NCD4aqyMoH9izuSjqNojAcHMpi-FaJg+NYd1LzqEwNMappW1YSBuAQEIgA)

Reply:
Hi @jakebailey, I understand the reasoning behind adding `| undefined` for consistency, but is it necessary in this context? My understanding is that when `exactOptionalPropertyTypes` is **enabled**, assigning `undefined` directly to optional properties can blur the distinction between a property being deliberately `undefined` and it being absent from the object. Should the implementation details of a technology (e.g., TypeScript) influence the representation of intent, such as what is necessary to connect with Vault?

For connectivity configurations to HashiCorp Vault, it would be more semantically correct to have properties like secret_id simply not present, as it reflects the absence of configuration data rather than an 'undefined' value being set. Using the absence of a property would make the intent of the code clearer and reduce potential ambiguity in type-checking.

Could you outline any specific scenarios in which explicitly setting these properties to `undefined` would be advantageous for Vault connectivity?

Reply:
Consider:

```ts
declare function doSomething(opts: AppRoleAuthConfig): void;

const secretId = foo ? "test" : undefined;

doSomething({ role_id: "role", secret_id: secretId });
```

This code would not compile under exactOptionalPropertyTypes without `| undefined`, even though it's likely it works.

Maybe it doesn't work if the library cares about the difference, in which case the extra type could be left off.

Reply:
Thank you @jakebailey - types updated where applicable